File: backend/internal/middleware/chat_rate_limit.go
package middleware

import (
	"time"

	"nexia-backend/internal/config"

	"github.com/gin-gonic/gin"
	"go.uber.org/zap"
)

const (
	defaultChatRateLimitRequests = 10
	defaultChatRateLimitBurst    = 3
	defaultChatRateLimitWindow   = time.Minute
)

// NewChatRateLimit returns an in-memory per-IP token-bucket limiter for chat endpoints.
// Requests that exceed the configured budget are rejected with HTTP 429.
func NewChatRateLimit(logger *zap.Logger, cfg *config.Config) gin.HandlerFunc {
	return newRateLimit(logger, "chat", "Too many chat requests", chatRateLimitConfigFromConfig(cfg))
}

func chatRateLimitConfigFromConfig(cfg *config.Config) rateLimitConfig {
	if cfg == nil {
		return rateLimitConfigFromValues(defaultChatRateLimitRequests, defaultChatRateLimitBurst, defaultChatRateLimitWindow, 0, 0, 0)
	}

	return rateLimitConfigFromValues(
		defaultChatRateLimitRequests,
		defaultChatRateLimitBurst,
		defaultChatRateLimitWindow,
		cfg.Server.ChatRateLimitRequests,
		cfg.Server.ChatRateLimitBurst,
		cfg.Server.ChatRateLimitWindowSeconds,
	)
}

File: apps/api/src/middleware/chat-rate-limit.ts
import type { Config } from "../config/config";
import type { Logger } from "../logging/logger";
import type { MiddlewareHandler } from "hono";
import { createRateLimiter, rateLimitConfigFromValues, type RateLimitConfig } from "./rate-limit";

const DEFAULT_CHAT_REQUESTS = 10;
const DEFAULT_CHAT_BURST = 3;
const DEFAULT_CHAT_WINDOW_SECONDS = 60;

function chatRateLimitConfigFromConfig(cfg: Config | null): RateLimitConfig {
  if (!cfg) {
    return rateLimitConfigFromValues(
      DEFAULT_CHAT_REQUESTS,
      DEFAULT_CHAT_BURST,
      DEFAULT_CHAT_WINDOW_SECONDS,
      0, 0, 0,
    );
  }
  return rateLimitConfigFromValues(
    DEFAULT_CHAT_REQUESTS,
    DEFAULT_CHAT_BURST,
    DEFAULT_CHAT_WINDOW_SECONDS,
    cfg.server.chat_rate_limit_requests,
    cfg.server.chat_rate_limit_burst,
    cfg.server.chat_rate_limit_window_seconds,
  );
}

export function createChatRateLimiter(logger: Logger, cfg: Config): MiddlewareHandler {
  const rlCfg = chatRateLimitConfigFromConfig(cfg);
  return createRateLimiter(logger, "chat", "Too many chat requests", rlCfg, "/chat");
}
